src/Workspaces/Core/Portable/Utilities/NoMainThreadDependencyAttribute.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.Utilities
+{
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Method | AttributeTargets.Parameter | AttributeTargets.Property | AttributeTargets.ReturnValue, AllowMultiple = false, Inherited = true)]
+    internal sealed class NoMainThreadDependencyAttribute : Attribute
```

Comment:
This seems like a reasonable solution to have the analyzer recognize and respect this attribute.
Do you want to add a `[Conditional]` attribute to this attribute so that it is dropped from your metadata? Or is the idea it propagates across metadata references so external consumers also benefit?

Reply:
> Or is the idea it propagates across metadata references so external consumers also benefit?

This is my current thought. However, it a branch locally I fanned this out a bit and there is a concerning number of cases to handle - and currently that's blindly assuming there are no calls to `JoinableTaskFactory.Run` from within a synchronous method.